pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java

```diff
@@ -149,7 +150,23 @@ public BaseDataBlock.Type getType() {
    * @return whether this block is the end of stream.
    */
   public boolean isEndOfStreamBlock() {
-    return _type == BaseDataBlock.Type.METADATA;
+    if (_isErrorBlock) {
+      return true;
+    } else if (_type != BaseDataBlock.Type.METADATA) {
+      return false;
+    }
+
+    MetadataBlock metadata = (MetadataBlock) _dataBlock;
+    return metadata.getType() == MetadataBlock.MetadataBlockType.EOS;
+  }
+
+  public boolean isNoOpBlock() {
+    if (_type != BaseDataBlock.Type.METADATA) {
+      return false;
+    }
+
+    MetadataBlock metadata = (MetadataBlock) _dataBlock;
+    return metadata.getType() == MetadataBlock.MetadataBlockType.NOOP;
   }
 
   /**
```

Comment:
let's also change the `isErrorBlock()`, no longer need a special boolean flag. since it is saved in the metadata block anyway